Reapply "[clang][CodeGen] Zero init unspecified fields in initializers in C" (#109898)

yabinc opened the pull request:
This reverts commit d50eaac12f0cdfe27e942290942b06889ab12a8c. Also fixes a bug calculating offsets for bit fields in the original patch.

yabinc committed:
Reapply "[clang][CodeGen] Zero init unspecified fields in initializers in C" (#109898)

This reverts commit d50eaac12f0cdfe27e942290942b06889ab12a8c.
Also fixes a bug calculating offsets for bit fields in the original patch.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1748,6 +1771,26 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+        uint64_t EndBitOffset = StartBitOffset + Info.Size;
+        SizeSoFar = CGF.getContext().toCharUnitsFromBits(EndBitOffset);
+        if (EndBitOffset % CGF.getContext().getCharWidth() != 0) {
+          SizeSoFar++;
```
I'm not sure this actually initializes everything that needs to be initialized?  If you end up with padding between bitfields for whatever reason, nothing actually writes those bits, I think.

yabinc replied:
I agree. I ignored padding less than 1 byte. It should be fixed in the new commit.
I think this problem doesn't exist for CGExprConstant.cpp. Because in const values, having a value for any bit of a byte can prevent this byte becoming undef value.

yabinc committed:
Zero init padding between bitfields in CGExprAgg.cpp

This isn't needed for CGExprConstant.cpp. Because an initialization
in one bit can prevent generating undef field in the const value.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1774,6 +1805,50 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+  };
+  auto InitBits = [&](uint64_t Byte, uint64_t Start, uint64_t End) {
+    Address Addr = CGF.Builder.CreateConstGEP(BaseAddr, Byte);
+    llvm::Value *Val = Builder.CreateLoad(Addr);
```
What's the size of the loaded integer here?  I can't find anything that explicitly sets it.

efriedma-quic replied:
(Ideally, we'd use load/store operations with the same width as the bitfield, as computed by CGBitFieldInfo, to make the result easier for LLVM to optimize.)

yabinc replied:
Originally it was Int8Ty. Now changed to store 0 to Info.StorageSize before initializing bitfields. I tested in -O2 that the store was always merged with initialization for bitfields.
Not sure if I need to consider Info.VolatileStorageSize, but I ran check-clang and test-suite successfully.

yabinc committed:
Init paddings between bitfields by storing 0 to StorageSize

When seeing a bitfield, store 0 to its [StorageOffset,
StorageOffset+StorageSize]. The store width is StorageSize,
making it easier to be merged with other bitfield stores.
And we don't need to worry about big/little endian problem.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1774,6 +1795,54 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+    CGF.Builder.CreateMemSet(Addr, Builder.getInt8(0), SizeVal, false);
+  };
+
+  if (NextField != nullptr && NextField->isBitField()) {
```
I had to spend a little time checking this actually works... but I think you're fine here: we should do initialization in field order, and any out-of-order access is undefined behavior.

efriedma-quic reviewed (approved):
LGTM
